[fix] Fixes related to block locators.

acoglio opened the pull request:
Rectify documentation of `BlockLocators` about validation during serialization. Also introduce more explicit notion of well-formedness, since subsequent documentation refers to that.

Make `BlockLocators::new_unchecked()` constructor, which does not perform validation, non-public, leveraging language-level protection. Also document that it is only used in tests.

As discussed on Slack, fix the check that the checkpoints map cover as many blocks as possible. Also explain the check.

As discussed on Slack, fix the check about overlapping checkpoints and recents. Also explain the check.

Remove a redundant check to validate block locators in `BlockSync`, as discussed on Slack, since deserialization already checks that.

Add TODOs related to either generalizing some of the code to `RECENT_INTERVAL > 1`, or removing such TODOs if we decide to hardwire `RECENT_INTERVAL` to 1 and eliminate that constant.

acoglio committed:
[fix] Fixes related to block locators.

Rectify documentation of `BlockLocators` about validation during serialization.
Also introduce more explicit notion of well-formedness, since subsequent
documentation refers to that.

Make `BlockLocators::new_unchecked()` constructor, which does not perform
validation, non-public, leveraging language-level protection. Also document that
it is only used in tests.

As discussed on Slack, fix the check that the checkpoints map cover as many
blocks as possible. Also explain the check.

As discussed on Slack, fix the check about overlapping checkpoints and recents.
Also explain the check.

Remove a redundant check to validate block locators in `BlockSync`, as discussed
on Slack, since deserialization already checks that.

Add TODOs related to either generalizing some of the code to `RECENT_INTERVAL >
1`, or removing such TODOs if we decide to hardwire `RECENT_INTERVAL` to 1 and
eliminate that constant.

acoglio committed:
Merge 'staging' and resolve conflicts.

acoglio committed:
Eliminate clippy dead-code error.

acoglio commented:
Closes #2185.

kaimast commented on node/sync/locators/src/block_locators.rs:
```diff
@@ -86,7 +88,9 @@ impl<N: Network> BlockLocators<N> {
     /// Initializes a new instance of the block locators, without checking the validity of the block locators.
-    pub fn new_unchecked(recents: IndexMap<u32, N::BlockHash>, checkpoints: IndexMap<u32, N::BlockHash>) -> Self {
+    /// This is only used for testing; note that it is non-public.
+    #[allow(dead_code)]
```
It is better to use `#[cfg(test)]` here.

acoglio replied:
Done.

kaimast commented on node/sync/locators/src/block_locators.rs:
```diff
@@ -189,27 +193,38 @@ impl<N: Network> BlockLocators<N> {
+        // All of this only works if `NUM_RECENT_BLOCKS < CHECKPOINT_INTERVAL`,
+        // because it is only under this condition that there is at most one overlapping height.
+        // TODO: generalize check for RECENT_INTERVAL > 1, or remove this comment if we hardwire that to 1
+        let last_recent_to_last_checkpoint_distance = last_recent_height % CHECKPOINT_INTERVAL;
```
Can we add a unit test where recents and checkpoints overlap?

acoglio replied:
[This](https://github.com/ProvableHQ/snarkOS/blob/4d4805796036709c367add54e3ab31f5900fa196/node/sync/locators/src/block_locators.rs#L531-L536) is already checking various situations with overlap.

kaimast commented on node/sync/locators/src/block_locators.rs:
```diff
@@ -189,27 +193,38 @@ impl<N: Network> BlockLocators<N> {
+        // must be greater than `last_recent_height`.
+        // We do not worry about this addition overflowing,
+        // because our use of `u32` for block heights implies that we do not expect to run out.
+        if last_checkpoint_height + CHECKPOINT_INTERVAL <= last_recent_height {
```
Shouldn't we also check that last_checkpoint_height is less or equal to last_recent_height, or is that already checked somewhere else?

Maybe we could do the following:
```rust
if !(last_checkpoint_height..last_checkpoint_height+CHECKPOINT_INTERVAL).contains(last_recent_height) {
       bail!(..);
}
```

acoglio replied:
Yes, that check was missing. Independently from that, I also realized that, since last_checkpoint_height is untrusted, we need to use a saturating add. So I used your nice suggested code, with a saturating add.

kaimast commented on node/sync/locators/src/block_locators.rs:
```diff
@@ -248,10 +263,12 @@ impl<N: Network> BlockLocators<N> {
 
         // If the last height is below NUM_RECENT_BLOCKS, ensure the number of recent blocks matches the last height.
+        // TODO: generalize check for RECENT_INTERVAL > 1, or remove this comment if we hardwire that to 1
         if last_height < NUM_RECENT_BLOCKS as u32 && recents.len().saturating_sub(1) as u32 != last_height {
```
I think we can just remove these checks and the TODO?

We already know that height increments by `RECENT_INTERVAL` (line 259) and that if `recents.len() < NUM_RECENT_BLOCKS` it holds that the first entry has height 0 (line 253).

From that, assuming `RECENT_INTERVAL` is positive, we can guarantee that either`recents.len() > NUM_RECENT_BLOCKS`, in which case `last_height >= NUM_RECENT_BLOCKS`, or `last_height == (recents.len()-1) * RECENT_INTERVAL`.

acoglio replied:
I think you are right. I've eliminated the check, and documented that it holds.

kaimast commented on node/sync/locators/src/block_locators.rs:
```diff
@@ -248,10 +263,12 @@ impl<N: Network> BlockLocators<N> {
         }
```
In line 253, could we write `*current_height > 0` instead of `*current_height != last_height`? The error message makes it clear that the assumption is that `last_height` is 0 at this point, which is indeed the case.

acoglio replied:
I was minimizing changes, but yes, that makes the code clearer and simpler. Pushed.

kaimast commented on node/sync/src/block_sync.rs:
```diff
@@ -420,15 +421,13 @@ impl<N: Network> BlockSync<N> {
             return Ok(());
         }
 
-        // Ensure the given block locators are well-formed.
```
Why was this removed? Maybe you can leave a comment here that states why we do not need to check for the validity of the block locators.

acoglio replied:
Yes, I updated the doc comment for the function.

kaimast reviewed (changes requested):
Looks good to me, aside from the minor notes in my review.

We should add a unit test, though. Ideally, something that would have failed before this fix. Considering there are already unit tests for this struct, adding another test should not take long.

acoglio committed:
Take suggestion from review.

acoglio committed:
Clarify code as suggested in review.

acoglio committed:
Add a clarifying comment, based on review.

acoglio committed:
Eliminate redundant checks uncovered in review.

And document the redundancy.

acoglio committed:
Add missing check and protect from overflow.

acoglio committed:
Fix off-by-one error in testing code.

acoglio commented:
The failing CI was due to an off-by-one error in the testing code itself. It was creating a BlockLocators structure with checkpoints map [0, 10000] and recents map [9900, ..., 9999], which is invalid, according to the recently added missing test.

acoglio committed:
Add unit test as suggested in review.

This failed before the fixes in the PR that this commit is part of, specifically
in commit 4d4805796 for example.

acoglio commented:
> We should add a unit test, though. Ideally, something that would have failed before this fix. Considering there are already unit tests for this struct, adding another test should not take long.

@kaimast I've added such a unit test.

acoglio committed:
Use unchecked constructor in test.

Since validity is checked right after, so that must have been the intent, also
for uniformity with other tests.

acoglio committed:
Document some testing functions better.

Remove a redundant check in a testing function, adding a comment to explain why
the check is redundant.

acoglio commented:
I've also made some improvements to the testing code (the last two commits).

raychu86 commented on node/sync/locators/src/block_locators.rs:
```diff
@@ -189,27 +193,44 @@ impl<N: Network> BlockLocators<N> {
         {
-            bail!("Recent genesis hash and checkpoint genesis hash mismatch at height {last_recent_height}")
+            bail!(
+                "Last checkpoint height ({last_checkpoint_height}) is not the largest multiple of ".to_string()
+                    + "{CHECKPOINT_INTERVAL} that does not exceed the last recent height ({last_recent_height})"
```
nit: Why is a string concatenation required here?

```suggestion
                "Last checkpoint height ({last_checkpoint_height}) is not the largest multiple of {CHECKPOINT_INTERVAL} that does not exceed the last recent height ({last_recent_height})"
```

acoglio replied:
I thought we might get a clippy complaint for exceeding 120 columns?

raychu86 replied:
You can make it more concise w/ something like:
``` 
bail!("Invalid checkpoint ({last_checkpoint_height}): must be the largest multiple of {CHECKPOINT_INTERVAL} ≤ {last_recent_height}");
```

Should be fine as long as clippy does not complain. Alternatively you can use the "\" character to split the lines. 

For example:
```
bail!(
    "Checkpoint height ({last_checkpoint_height}) must be the largest multiple of \
     {CHECKPOINT_INTERVAL} not exceeding recent height ({last_recent_height})"
);
```

acoglio replied:
Yes, I should have used `\`. Pushed.

raychu86 commented on node/sync/locators/src/block_locators.rs:
```diff
@@ -189,27 +193,44 @@ impl<N: Network> BlockLocators<N> {
+        // and thus we use a saturating addition;
+        // only a faulty validator would send block locators with such high block heights,
+        // under the assumption that the blockchain is always well below the `u32` limit for heights.
+        if !(last_checkpoint_height..last_checkpoint_height.saturating_add(CHECKPOINT_INTERVAL))
+            .contains(&last_recent_height)
```
Rust compiler may optimize this away, but why do a `contains` check in a range rather than 2 comparator checks?

acoglio replied:
I followed a suggestion of another review. Either way is fine with me, so long as there's agreement.

raychu86 commented:
Left a few comments, mainly stylistic changes. Changes to the checks seem to make sense, but I would like to have this tested further.

@kpandl Could you run the sync tests on this branch to ensure there are no regressions or unwanted behavior?

acoglio committed:
Take review suggestion.

raychu86 reviewed (commented):
Logic looks good, but do not merge until we have confirmation of sync tests

kpandl reviewed (approved):
LGTM, syncing works [[Grafana](https://aleostresstest.grafana.net/d/single-region-tests-kp/single-region-tests-kp?from=2025-03-21T12:43:18.054Z&to=2025-03-21T13:06:05.035Z&refresh=5m&timezone=browser&orgId=1)]
